ENH: Builder saving messages to hdf5 and other related changes

isolver opened the pull request:
ENH: If iohub is running and saving a hdf5 file, Builder now saves visual stim start time messages to hdf5
ENH: Added `saveEventReport` function to `psychopy.iohub.datastore.util`
BF: P3 str / bytes issues in `psychopy.iohub.datastore.util`
ENH: Added iohub hdf5 reader to gcCursor demo that uses experiment conditions table
RF: Switched datastore coder demo to use new `saveEventReport` function
ENH: Added `cacheMessageEvents()` and `sendMessageEvents()` and to iohub connection class. `sendMessageEvents()` also sends any cached messages.

isolver committed:
BF: Fixed bytes != str issue in iohub datastore util class

BF: Fixed bytes != str issue in iohub datastore util class

isolver committed:
ADD: include hdf5 sample parsing script for gcCursor demo

ADD: include hdf5 sample parser for gcCursor demo. samples are split into trials using the datastore condition variable table. Trial's condition variable value for each sample is written along with sample data.

isolver committed:
ADD: Include hdf5 parsing scripts

ADD: Include eye sample parsing scripts using either experiment messages or the datastore trial condition variables table to split samples into trials.
RF: Update trial start / stop messages written by gcCursor demo

isolver committed:
RF: Added some comments to readTrialSamplesByMessages

RF: Added some comments to readTrialSamplesByMessages

isolver committed:
RF: pep8 formatting changes to new hdf5 file reading demos

RF: pep8 formatting changes to new hdf5 file reading demos

isolver committed:
ADD: demo for saving hdf5 binocular samples to txt using experiment msgs

ADD: demo for saving hdf5 binocular samples to txt using experiment msgs

isolver committed:
RF: Change new examples to (possibly) handle any event type

RF: Monocular and Binocular eye sample event types are included in examples

isolver committed:
FF: Disable ability to store multiple sessions in one hdf5 file

FF: Disable ability to store multiple sessions in one hdf5 file

isolver committed:
RF: some cleanup of iohub.datastore source

RF: some cleanup of iohub.datastore source

isolver committed:
RF: Demo description cleanup

RF: Demo description cleanup

isolver committed:
WIP: ENH: Add hdf5 onset and offset messages to Builder visual components.

WIP: ENH: Add hdf5 onset and offset messages to Builder visual components.

isolver committed:
ENH: iohub multipacket request support

ENH: iohub multipacket request support

isolver committed:
FF: remove onset and offsetMessage params from base visual component

FF: remove onset and offsetMessage params from base visual component
FF: use auto generated iohub exp message text == component name + '_onset' or + '_offset'

isolver committed:
Merge remote-tracking branch 'upstream/dev' into enh_builder_iohub_messages

isolver committed:
Merge remote-tracking branch 'upstream/dev' into iohub_hdf5_to_text

isolver committed:
Merge branch 'iohub_hdf5_to_text' into enh_builder_iohub_messages

isolver committed:
RF: Remove print from mousegaze calibration code

isolver committed:
FF: Send cached experiment messages before shutting down iohub server

isolver committed:
FF: Builder generated iohub messages use same labels as psychopy results file

FF: Builder generated iohub messages use componentLabel+'.started' or '.stopped' so they match column names saved to standard psychopy results file

isolver committed:
WIP: Add hdf5 conversion script for Builder demos

WIP: Add hdf5 conversion script for Builder eye tracking demos, should eventually move most to iohub.datastore.util so demo only needs to be a dozen lines of code or so.

isolver committed:
FF: More meaningful stim name

FF: More meaningful stim name

isolver committed:
ENH: added saveEventReport function to iohub.datastore.util

ENH: added saveEventReport function to iohub.datastore.util
RF: rename iohub datastore coder demo and use iohub.datastore.util.saveEventReport
BF: displayEventTableSelectionDialog was not using list_label arg

isolver committed:
FF: convert bytes to str and use '.' for missing str value

FF: convert bytes to str and use '.' for missing str value

isolver committed:
RF: use datastore.util.saveEventReport in gcCursor hdf5 parser

RF: use datastore.util.saveEventReport in gcCursor hdf5 parser

isolver committed:
RF: Remove redundant iohub datastore demo

RF: Remove redundant iohub datastore demo

isolver committed:
FF: Only save component start times to hdf5 file

FF: Only save visual component start times to hdf5 file

isolver committed:
FF: Added timeMargins arg to saveEventReport, default of [0.0, 0.0]

FF: Added timeMargins arg to saveEventReport, default of [0.0, 0.0]. Can be used to expand trial period beyond start / stop message times

isolver committed:
Merge remote-tracking branch 'upstream/dev' into enh_builder_iohub_messages

isolver committed:
RF: lgtm fixes

TEParsons commented on psychopy/experiment/components/_base.py:
```diff
@@ -808,6 +809,29 @@ def writeFrameCodeJS(self, buff):
+                if self.params['syncScreenRefresh'].val:
+                        code = (
+                            f"if ioServer:\n"
+                            f"    ioServer.cacheMessageEvent('{comp_onset_tag}', sec_time={name}.tStartRefresh)\n"
```
Adding tabs to boilerplate like this can cause an indentation error in some cases, it's safer to do it like this:
```suggestion
                code = (
                        f"if ioServer:\n"
                )
                buff.writeIndentedLines(code)
                buff.setIndentLevel(1, relative=True)
                if self.params['syncScreenRefresh'].val:
                        code = (
                                f"ioServer.cacheMessageEvent('{comp_onset_tag}', sec_time={name}.tStartRefresh)\n"
                        )
                else:
                        code = (
                                f"ioServer.cacheMessageEvent('{comp_onset_tag}', sec_time={name}.tStart)\n"
                        )
                buff.writeIndentedLines(code)
                buff.setIndentLevel(-1, relative=True)
```
as `buff.setIndentLevel` will figure out how much to indent by based on user's prefs rather than assuming an indent is 4 spaces.

(though there's loads of instances where we've already done indentation wrong, particularly in older components)

isolver committed:
Update psychopy/experiment/components/_base.py

Co-authored-by: Todd Parsons <todd.e.parsons@googlemail.com>

isolver committed:
FF: fix code after merge issue

isolver committed:
BF: iohub datastore  util use SESSION_ID instead of session_id

BF: iohub datastore  util use SESSION_ID instead of session_id, as some files maybe saving session_id as string.
